src/Tasks/TaskTranslatorHelpers.cs

```diff
@@ -34,5 +35,94 @@ public static void Translate(this ITranslator translator, ref FrameworkName fram
                 frameworkName = new FrameworkName(identifier, version, profile);
             }
         }
+
+        public static void TranslateDictionary(this ITranslator translator, ref Dictionary<string, DateTime> dict, StringComparer comparer)
+        {
+            int count = 0;
+            if (translator.Mode == TranslationDirection.ReadFromStream)
+            {
+                dict = new Dictionary<string, DateTime>(comparer);
+                translator.Translate(ref count);
+                string key = string.Empty;
+                DateTime val = DateTime.Now;
+                for (int i = 0; i < count; i++)
+                {
+                    translator.Translate(ref key);
+                    translator.Translate(ref val);
+                    dict.Add(key, val);
+                }
+            }
+            else
+            {
+                count = dict.Count;
+                translator.Translate(ref count);
+                foreach (KeyValuePair<string, DateTime> kvp in dict)
+                {
+                    string key = kvp.Key;
+                    DateTime val = kvp.Value;
+                    translator.Translate(ref key);
+                    translator.Translate(ref val);
+                }
+            }
+        }
+
+        public static void Translate(this ITranslator translator, ref Dependencies dependencies, Type t)
```

Comment:
Why are you putting these here instead of as a method in the relevant classes?

Reply:
When we've added new Translators, we've generally done it in "TranslatorHelpers" (or similar) files. I was thinking about eventually combining these with @rokonec's change for the other kind of StateFileBase: https://github.com/dotnet/msbuild/pull/6094

TaskTranslatorHelpers makes a little more sense to me than a shared TranslatorHelpers, though, because these are all task-specific.

Reply:
I believe that for for our own types we shall implement ITranslatable interface and use extension only for close source clases, like Dictionary<X,Y> or Version. 
This was the serialization code stays inside the class, which might limit futuru serialization code omission bugs.